File: orchestrator/apps/templates/workflow/executor.py
"""
DAG Executor for Workflow Engine.

Executes workflow DAG in topological order with:
- Topological ordering via DAGValidator
- Conditional edge evaluation (Jinja2 expressions)
- Node skipping based on conditions
- Error handling and rollback tracking
- Support for all node types (operation, condition, parallel, loop, subworkflow)
"""

import inspect
import logging
from typing import Any, Dict, List, Optional, Set, Tuple

from asgiref.sync import sync_to_async

from apps.templates.tracing import (
    add_span_event,
    get_current_span_id,
    set_span_attribute,
    set_span_error,
    start_node_span,
)
from apps.templates.consumers import (
    broadcast_node_update,
    broadcast_workflow_update,
)
from apps.templates.workflow.context import ContextManager
from apps.templates.workflow.handlers import NodeExecutionMode, NodeHandlerFactory
from apps.templates.workflow.models import (
    DAGStructure,
    WorkflowEdge,
    WorkflowExecution,
    WorkflowNode,
)
from apps.templates.workflow.validator import DAGValidator

logger = logging.getLogger(__name__)


class DAGExecutionError(Exception):
    """Exception raised during DAG execution."""

    def __init__(
        self,
        message: str,
        node_id: Optional[str] = None,
        recoverable: bool = False
    ):
        """
        Initialize DAG execution error.

        Args:
            message: Error description
            node_id: Node ID where error occurred (optional)
            recoverable: Whether error is recoverable (for retry logic)
        """
        self.message = message
        self.node_id = node_id
        self.recoverable = recoverable
        super().__init__(self.message)


class DAGExecutor:
    """
    Executes workflow DAG in topological order.

    Features:
    - Topological order execution (uses validator.topological_order)
    - Conditional edge evaluation (Jinja2 expressions)
    - Node skipping based on conditions
    - Error handling and rollback tracking
    - Integration with NodeHandlers for node execution

    Usage:
        executor = DAGExecutor(dag, execution)
        success, result = await executor.execute(context)
    """

    def __init__(self, dag: DAGStructure, execution: WorkflowExecution):
        """
        Initialize DAGExecutor with DAG and execution instance.

        Args:
            dag: DAG structure to execute
            execution: WorkflowExecution instance for state tracking

        Raises:
            ValueError: If DAG is invalid
        """
        self.dag = dag
        self.execution = execution

        # Build node map for quick lookup
        self.node_map: Dict[str, WorkflowNode] = {
            node.id: node for node in dag.nodes
        }

        # Build edge maps
        self.outgoing_edges: Dict[str, List[WorkflowEdge]] = {}
        self.incoming_edges: Dict[str, List[WorkflowEdge]] = {}

        for edge in dag.edges:
            # Outgoing edges (from_node -> [edges])
            if edge.from_node not in self.outgoing_edges:
                self.outgoing_edges[edge.from_node] = []
            self.outgoing_edges[edge.from_node].append(edge)

            # Incoming edges (to_node -> [edges])
            if edge.to_node not in self.incoming_edges:
                self.incoming_edges[edge.to_node] = []
            self.incoming_edges[edge.to_node].append(edge)

        # Get topological order
        self._topological_order: Optional[List[str]] = None
        self._validate_and_get_order()

        logger.info(
            f"DAGExecutor initialized for execution {execution.id}",
            extra={
                'execution_id': str(execution.id),
                'node_count': len(dag.nodes),
                'edge_count': len(dag.edges)
            }
        )

    def _validate_and_get_order(self) -> None:
        """
        Validate DAG and extract topological order.

        Raises:
            ValueError: If DAG validation fails
        """
        validator = DAGValidator(self.dag)
        result = validator.validate()

        if not result.is_valid:
            error_messages = [issue.message for issue in result.errors]
            raise ValueError(f"DAG validation failed: {'; '.join(error_messages)}")

        self._topological_order = result.topological_order

        logger.debug(
            "DAG validated successfully",
            extra={
                'topological_order': self._topological_order,
                'warnings': len(result.warnings)
            }
        )

    async def execute(self, context: ContextManager) -> Tuple[bool, Dict[str, Any]]:
        """
        Execute DAG in topological order.

        Iterates through nodes in topological order, executing each node
        that should be executed based on incoming edge conditions.

        Args:
            context: Initial execution context

        Returns:
            Tuple[success, final_context_or_error]:
                - (True, final_context_dict) if successful
                - (False, {'error': message, 'node_id': id}) if failed
        """
        if not self._topological_order:
            return False, {'error': 'No topological order available', 'node_id': None}

        current_context = context.snapshot()
        executed_nodes: Set[str] = set()
        skipped_nodes: Set[str] = set()

        logger.info(
            f"Starting DAG execution for {self.execution.id}",
            extra={
                'execution_id': str(self.execution.id),
                'node_order': self._topological_order
            }
        )

        try:
            for node_id in self._topological_order:
                node = self.node_map.get(node_id)
                if not node:
                    raise DAGExecutionError(
                        f"Node '{node_id}' not found in node map",
                        node_id=node_id
                    )

                # Check if node should be executed
                should_execute = self._should_execute_node(
                    node_id, current_context, executed_nodes, skipped_nodes
                )

                if not should_execute:
                    # Skip this node
                    skipped_nodes.add(node_id)
                    await sync_to_async(
                        self.execution.update_node_status,
                        thread_sensitive=True
                    )(node_id, 'skipped')

                    # Broadcast node skip via WebSocket
                    await self._broadcast_node_update(node_id, 'skipped')

                    logger.info(
                        f"Skipping node {node_id} (conditions not met)",
                        extra={
                            'execution_id': str(self.execution.id),
                            'node_id': node_id
                        }
                    )
                    continue

                # Execute node
                success, current_context = await self._execute_node(
                    node_id, node, current_context
                )

                if not success:
                    # Node execution failed
                    error_info = current_context.get('_last_error', {})
                    return False, {
                        'error': error_info.get('message', 'Node execution failed'),
                        'node_id': node_id,
                        'context': current_context.to_dict()
                    }

                executed_nodes.add(node_id)

            # All nodes executed successfully
            logger.info(
                f"DAG execution completed successfully for {self.execution.id}",
                extra={
                    'execution_id': str(self.execution.id),
                    'executed_count': len(executed_nodes),
                    'skipped_count': len(skipped_nodes)
                }
            )

            return True, current_context.to_dict()

        except DAGExecutionError as exc:
            logger.error(
                f"DAG execution failed: {exc.message}",
                extra={
                    'execution_id': str(self.execution.id),
                    'node_id': exc.node_id
                },
                exc_info=True
            )
            return False, {
                'error': exc.message,
                'node_id': exc.node_id,
                'recoverable': exc.recoverable
            }

        except Exception as exc:
            logger.error(
                f"Unexpected error during DAG execution: {exc}",
                extra={'execution_id': str(self.execution.id)},
                exc_info=True
            )
            return False, {
                'error': str(exc),
                'node_id': None,
                'recoverable': False
            }

    def _should_execute_node(
        self,
        node_id: str,
        context: ContextManager,
        executed_nodes: Set[str],
        skipped_nodes: Set[str]
    ) -> bool:
        """
        Check if node should execute based on incoming edge conditions.

        A node should execute if:
        1. It has no incoming edges (start node), OR
        2. At least one incoming edge has its condition satisfied AND
           the source node was executed (not skipped)

        Args:
            node_id: Node ID to check
            context: Current execution context
            executed_nodes: Set of already executed node IDs
            skipped_nodes: Set of skipped node IDs

        Returns:
            True if node should execute
        """
        # Get incoming edges for this node
        incoming = self.incoming_edges.get(node_id, [])

        # Start nodes (no incoming edges) always execute
        if not incoming:
            return True

        # Check each incoming edge
        for edge in incoming:
            source_node = edge.from_node

            # Source must have been executed (not skipped)
            if source_node in skipped_nodes:
                continue

            if source_node not in executed_nodes:
                # Source not yet processed - this shouldn't happen in topological order
                logger.warning(
                    f"Source node {source_node} not yet executed for edge to {node_id}",
                    extra={'execution_id': str(self.execution.id)}
                )
                continue

            # Check edge condition
            if self._evaluate_edge_condition(edge, context):
                return True

        # No valid incoming path found
        return False

    def _evaluate_edge_condition(
        self,
        edge: WorkflowEdge,
        context: ContextManager
    ) -> bool:
        """
        Evaluate edge condition expression.

        Args:
            edge: Edge with optional condition
            context: Execution context for variable resolution

        Returns:
            True if condition is satisfied (or no condition)
        """
        # No condition = always true
        if not edge.condition:
            return True

        try:
            result = context.evaluate_condition(edge.condition)

            logger.debug(
                f"Edge condition evaluated: {edge.from_node} -> {edge.to_node}",
                extra={
                    'condition': edge.condition,
                    'result': result
                }
            )

            return result

        except Exception as exc:
            logger.warning(
                f"Edge condition evaluation failed: {exc}",
                extra={
                    'from_node': edge.from_node,
                    'to_node': edge.to_node,
                    'condition': edge.condition
                }
            )
            # On error, treat as false (don't execute)
            return False

    async def _execute_node(
        self,
        node_id: str,
        node: WorkflowNode,
        context: ContextManager
    ) -> Tuple[bool, ContextManager]:
        """
        Execute single node and return updated context.

        Args:
            node_id: Node identifier
            node: WorkflowNode to execute
            context: Current execution context

        Returns:
            Tuple[success, updated_context]:
                - (True, context_with_result) if successful
                - (False, context_with_error) if failed
        """
        logger.info(
            f"Executing node {node_id} (type: {node.type})",
            extra={
                'execution_id': str(self.execution.id),
                'node_id': node_id,
                'node_type': node.type
            }
        )

        # Update node status to running
        await sync_to_async(
            self.execution.update_node_status,
            thread_sensitive=True
        )(node_id, 'running')

        # Broadcast node start via WebSocket
        await self._broadcast_node_update(node_id, 'running')

        # Start OpenTelemetry span for node execution
        with start_node_span(
            node_id=node_id,
            node_name=node.name,
            node_type=node.type,
            execution_id=str(self.execution.id)
        ) as span:
            try:
                # Get appropriate handler
                handler = NodeHandlerFactory.get_handler(node.type)

                if span:
                    add_span_event("node_handler_resolved", {"handler_type": node.type})

                # Execute node
                if inspect.iscoroutinefunction(handler.execute):
                    result = await handler.execute(
                        node=node,
                        context=context.to_dict(),
                        execution=self.execution,
                        mode=NodeExecutionMode.SYNC
                    )
                else:
                    result = await sync_to_async(
                        handler.execute,
                        # IMPORTANT: do not run long-running node handlers in the
                        # global "thread_sensitive" executor. Operation nodes may
                        # synchronously wait for Go Worker completion, and that
                        # would starve other thread_sensitive DB tasks (including
                        # WebSocket handshake/status queries), causing frequent
                        # WS disconnect/reconnect during workflow execution.
                        thread_sensitive=False
                    )(
                        node=node,
                        context=context.to_dict(),
                        execution=self.execution,
                        mode=NodeExecutionMode.SYNC
                    )

                if result.success:
                    # Update context with node result
                    updated_context = context.add_node_result(
                        node_id,
                        {
                            'success': True,
                            'output': result.output,
                            'duration_seconds': result.duration_seconds
                        }
                    )

                    # Update node status
                    await sync_to_async(
                        self.execution.update_node_status,
                        thread_sensitive=True
                    )(
                        node_id, 'completed',
                        result={'output': result.output, 'duration': result.duration_seconds}
                    )

                    logger.info(
                        f"Node {node_id} completed successfully",
                        extra={
                            'execution_id': str(self.execution.id),
                            'node_id': node_id,
                            'duration_seconds': result.duration_seconds
                        }
                    )

                    # Broadcast node completion via WebSocket
                    await self._broadcast_node_update(
                        node_id, 'completed',
                        output=result.output,
                        duration_ms=int((result.duration_seconds or 0) * 1000)
                    )

                    if span:
                        set_span_attribute("node.status", "completed")
                        set_span_attribute("node.duration_seconds", result.duration_seconds)
                        add_span_event("node_completed")

                    return True, updated_context

                else:
                    # Node execution failed
                    error_context = context.add_node_result(
                        node_id,
                        {
                            'success': False,
                            'error': result.error,
                            'duration_seconds': result.duration_seconds
                        }
                    )

                    # Store last error for reporting
                    error_context = error_context.set('_last_error', {
                        'node_id': node_id,
                        'message': result.error
                    })

                    # Update node status
                    await sync_to_async(
                        self.execution.update_node_status,
                        thread_sensitive=True
                    )(
                        node_id, 'failed',
                        result={'error': result.error, 'duration': result.duration_seconds}
                    )

                    logger.error(
                        f"Node {node_id} failed: {result.error}",
                        extra={
                            'execution_id': str(self.execution.id),
                            'node_id': node_id,
                            'error': result.error
                        }
                    )

                    # Broadcast node failure via WebSocket
                    await self._broadcast_node_update(
                        node_id, 'failed',
                        error=result.error,
                        duration_ms=int((result.duration_seconds or 0) * 1000)
                    )

                    if span:
                        set_span_attribute("node.status", "failed")
                        set_span_attribute("node.error", result.error)
                        add_span_event("node_failed", {"error": result.error})

                    return False, error_context

            except ValueError as exc:
                # Handler not found or invalid
                error_msg = f"Handler error for node {node_id}: {exc}"
                logger.error(error_msg, extra={'node_id': node_id}, exc_info=True)

                await sync_to_async(
                    self.execution.update_node_status,
                    thread_sensitive=True
                )(
                    node_id, 'failed',
                    result={'error': error_msg}
                )

                error_context = context.set('_last_error', {
                    'node_id': node_id,
                    'message': error_msg
                })

                if span:
                    set_span_error(exc)

                return False, error_context

            except Exception as exc:
                # Unexpected error
                error_msg = f"Unexpected error executing node {node_id}: {exc}"
                logger.error(error_msg, extra={'node_id': node_id}, exc_info=True)

                await sync_to_async(
                    self.execution.update_node_status,
                    thread_sensitive=True
                )(
                    node_id, 'failed',
                    result={'error': error_msg}
                )

                error_context = context.set('_last_error', {
                    'node_id': node_id,
                    'message': error_msg
                })

                if span:
                    set_span_error(exc)

                return False, error_context

    def get_next_nodes(
        self,
        current_node_id: str,
        context: ContextManager
    ) -> List[str]:
        """
        Get next nodes to execute based on edge conditions.

        Used for async execution to determine next steps.

        Args:
            current_node_id: Current node ID
            context: Current execution context

        Returns:
            List of next node IDs that should be executed
        """
        next_nodes = []

        # Get outgoing edges
        outgoing = self.outgoing_edges.get(current_node_id, [])

        for edge in outgoing:
            if self._evaluate_edge_condition(edge, context):
                next_nodes.append(edge.to_node)

        logger.debug(
            f"Next nodes for {current_node_id}: {next_nodes}",
            extra={
                'execution_id': str(self.execution.id),
                'current_node': current_node_id,
                'next_nodes': next_nodes
            }
        )

        return next_nodes

    async def _broadcast_node_update(
        self,
        node_id: str,
        status: str,
        output: Optional[Dict[str, Any]] = None,
        error: Optional[str] = None,
        duration_ms: Optional[int] = None
    ) -> None:
        """
        Broadcast node status update via WebSocket.

        Args:
            node_id: Node identifier
            status: New node status (running, completed, failed, skipped)
            output: Optional output data (for completed)
            error: Optional error message (for failed)
            duration_ms: Optional duration in milliseconds
        """
        try:
            # Calculate progress based on completed nodes
            total_nodes = len(self._topological_order) if self._topological_order else 1
            node_statuses = self.execution.node_statuses or {}

            # Count completed and failed nodes
            completed = sum(
                1 for s in node_statuses.values()
                if isinstance(s, dict) and s.get('status') in ('completed', 'failed', 'skipped')
            )

            progress = completed / total_nodes if total_nodes > 0 else 0

            # Get current trace context
            trace_id = self.execution.trace_id
            span_id = get_current_span_id()

            # Broadcast node update
            await broadcast_node_update(
                execution_id=str(self.execution.id),
                node_id=node_id,
                status=status,
                output=output,
                error=error,
                duration_ms=duration_ms,
                span_id=span_id
            )

            # Also broadcast workflow progress update
            await broadcast_workflow_update(
                execution_id=str(self.execution.id),
                status='running',
                progress=progress,
                current_node_id=node_id,
                trace_id=trace_id
            )

        except Exception as exc:
            # Don't fail execution if broadcast fails
            logger.warning(
                f"Failed to broadcast node update: {exc}",
                extra={
                    'execution_id': str(self.execution.id),
                    'node_id': node_id,
                    'status': status
                }
            )

    @property
    def topological_order(self) -> List[str]:
        """Get topological order of nodes."""
        return self._topological_order or []

File: go-services/worker/internal/workflow/executor/executor.go
// Package executor provides DAG execution engine for workflow orchestration.
package executor

import (
	"context"
	"fmt"
	"sort"
	"time"

	"go.uber.org/zap"

	"github.com/commandcenter1c/commandcenter/worker/internal/workflow"
	wfcontext "github.com/commandcenter1c/commandcenter/worker/internal/workflow/context"
	"github.com/commandcenter1c/commandcenter/worker/internal/workflow/models"
	"github.com/commandcenter1c/commandcenter/worker/internal/workflow/validator"
)

// NodeStatus represents the execution status of a node.
type NodeStatus string

const (
	// NodeStatusPending indicates the node has not started.
	NodeStatusPending NodeStatus = "pending"
	// NodeStatusRunning indicates the node is currently executing.
	NodeStatusRunning NodeStatus = "running"
	// NodeStatusCompleted indicates the node completed successfully.
	NodeStatusCompleted NodeStatus = "completed"
	// NodeStatusFailed indicates the node execution failed.
	NodeStatusFailed NodeStatus = "failed"
	// NodeStatusSkipped indicates the node was skipped due to conditions.
	NodeStatusSkipped NodeStatus = "skipped"
)

// NodeResult represents the result of executing a single node.
type NodeResult struct {
	// NodeID is the identifier of the executed node.
	NodeID string
	// Status is the execution status.
	Status NodeStatus
	// Output is the node's output data.
	Output interface{}
	// Error is the error if execution failed.
	Error error
	// Duration is how long execution took.
	Duration time.Duration
	// StartedAt is when execution started.
	StartedAt time.Time
	// CompletedAt is when execution completed.
	CompletedAt time.Time
}

// NodeHandler interface for executing nodes.
// Implementations handle specific node types (operation, condition, parallel, etc.).
type NodeHandler interface {
	// HandleNode executes a node and returns the result.
	HandleNode(ctx context.Context, node *models.Node, execCtx *wfcontext.ExecutionContext) (*NodeResult, error)
	// SupportedTypes returns the node types this handler can process.
	SupportedTypes() []models.NodeType
}

// ConditionEvaluator interface for evaluating edge conditions.
type ConditionEvaluator interface {
	// Evaluate evaluates a condition expression against the context.
	// Returns true if condition is satisfied, false otherwise.
	Evaluate(condition string, ctx *wfcontext.ExecutionContext) (bool, error)
}

// ExecutionCallback is called during execution for status updates.
type ExecutionCallback func(event ExecutionEvent)

// ExecutionEvent represents an event during execution.
type ExecutionEvent struct {
	Type      EventType
	NodeID    string
	Status    NodeStatus
	Result    *NodeResult
	Progress  float64
	Error     error
	Timestamp time.Time
}

// EventType classifies execution events.
type EventType string

const (
	EventNodeStarted   EventType = "node_started"
	EventNodeCompleted EventType = "node_completed"
	EventNodeFailed    EventType = "node_failed"
	EventNodeSkipped   EventType = "node_skipped"
	EventProgress      EventType = "progress"
)

// ExecutorConfig holds configuration for the executor.
type ExecutorConfig struct {
	// MaxConcurrent is the maximum number of concurrent node executions (for parallel nodes).
	MaxConcurrent int
	// DefaultTimeout is the default node execution timeout.
	DefaultTimeout time.Duration
	// StopOnError stops execution on first node failure if true.
	StopOnError bool
}

// DefaultExecutorConfig returns sensible defaults.
func DefaultExecutorConfig() *ExecutorConfig {
	return &ExecutorConfig{
		MaxConcurrent:  10,
		DefaultTimeout: 5 * time.Minute,
		StopOnError:    true,
	}
}

// Executor executes a workflow DAG in topological order.
type Executor struct {
	dag              *models.DAG
	handlers         map[models.NodeType]NodeHandler
	conditionEval    ConditionEvaluator
	config           *ExecutorConfig
	logger           *zap.Logger
	callback         ExecutionCallback
	topologicalOrder []string
	incomingEdges    map[string][]*models.Edge
	outgoingEdges    map[string][]*models.Edge
}

// NewExecutor creates a new DAG executor.
func NewExecutor(dag *models.DAG, logger *zap.Logger) (*Executor, error) {
	return NewExecutorWithConfig(dag, logger, DefaultExecutorConfig())
}

// NewExecutorWithConfig creates a new DAG executor with custom configuration.
func NewExecutorWithConfig(dag *models.DAG, logger *zap.Logger, config *ExecutorConfig) (*Executor, error) {
	if dag == nil {
		return nil, workflow.ErrEmptyDAG
	}
	if logger == nil {
		logger = zap.NewNop()
	}
	if config == nil {
		config = DefaultExecutorConfig()
	}

	// Validate DAG and get topological order
	v := validator.NewValidator(dag)
	result := v.Validate()
	if !result.IsValid {
		errMsg := "DAG validation failed"
		if len(result.Errors) > 0 {
			errMsg = result.Errors[0].Message
		}
		return nil, workflow.NewValidationError(errMsg)
	}

	// Build edge maps
	incomingEdges := make(map[string][]*models.Edge)
	outgoingEdges := make(map[string][]*models.Edge)

	for _, edge := range dag.Edges {
		incomingEdges[edge.To] = append(incomingEdges[edge.To], edge)
		outgoingEdges[edge.From] = append(outgoingEdges[edge.From], edge)
	}

	e := &Executor{
		dag:              dag,
		handlers:         make(map[models.NodeType]NodeHandler),
		config:           config,
		logger:           logger,
		topologicalOrder: result.TopologicalOrder,
		incomingEdges:    incomingEdges,
		outgoingEdges:    outgoingEdges,
	}

	logger.Debug("Executor initialized",
		zap.String("dag_id", dag.ID),
		zap.Int("node_count", len(dag.Nodes)),
		zap.Int("edge_count", len(dag.Edges)),
		zap.Strings("topological_order", result.TopologicalOrder))

	return e, nil
}

// RegisterHandler registers a node handler for the specified types.
func (e *Executor) RegisterHandler(handler NodeHandler) {
	for _, nodeType := range handler.SupportedTypes() {
		e.handlers[nodeType] = handler
	}
}

// SetConditionEvaluator sets the condition evaluator.
func (e *Executor) SetConditionEvaluator(eval ConditionEvaluator) {
	e.conditionEval = eval
}

// SetCallback sets the execution callback for status updates.
func (e *Executor) SetCallback(callback ExecutionCallback) {
	e.callback = callback
}

// TopologicalOrder returns the topologically sorted node IDs.
func (e *Executor) TopologicalOrder() []string {
	return e.topologicalOrder
}

// Execute runs the workflow and returns the final context or error.
func (e *Executor) Execute(ctx context.Context, execCtx *wfcontext.ExecutionContext) (*wfcontext.ExecutionContext, error) {
	if len(e.topologicalOrder) == 0 {
		return nil, workflow.NewValidationError("no topological order available")
	}

	currentCtx := execCtx.Clone()
	executedNodes := make(map[string]bool)
	skippedNodes := make(map[string]bool)
	nodeStatuses := make(map[string]NodeStatus)

	// Initialize all nodes as pending
	for _, nodeID := range e.topologicalOrder {
		nodeStatuses[nodeID] = NodeStatusPending
	}

	e.logger.Info("Starting DAG execution",
		zap.String("execution_id", execCtx.ExecutionID()),
		zap.Int("total_nodes", len(e.topologicalOrder)))

	totalNodes := len(e.topologicalOrder)
	completedCount := 0

	for _, nodeID := range e.topologicalOrder {
		// Check for cancellation
		select {
		case <-ctx.Done():
			return nil, workflow.NewCancelledError("execution cancelled")
		default:
		}

		node := e.dag.GetNode(nodeID)
		if node == nil {
			return nil, workflow.NewInvalidNodeError(nodeID, "node not found in DAG")
		}

		// Check if node should be executed based on incoming edges
		shouldExecute := e.shouldExecuteNode(nodeID, currentCtx, executedNodes, skippedNodes)

		if !shouldExecute {
			skippedNodes[nodeID] = true
			nodeStatuses[nodeID] = NodeStatusSkipped
			completedCount++

			e.logger.Debug("Skipping node (conditions not met)",
				zap.String("node_id", nodeID),
				zap.String("execution_id", execCtx.ExecutionID()))

			e.emitEvent(ExecutionEvent{
				Type:      EventNodeSkipped,
				NodeID:    nodeID,
				Status:    NodeStatusSkipped,
				Progress:  float64(completedCount) / float64(totalNodes),
				Timestamp: time.Now(),
			})
			continue
		}

		// Execute the node
		currentCtx.SetCurrentNode(nodeID)
		nodeStatuses[nodeID] = NodeStatusRunning

		e.emitEvent(ExecutionEvent{
			Type:      EventNodeStarted,
			NodeID:    nodeID,
			Status:    NodeStatusRunning,
			Timestamp: time.Now(),
		})

		e.logger.Debug("Executing node",
			zap.String("node_id", nodeID),
			zap.String("node_type", string(node.Type)),
			zap.String("execution_id", execCtx.ExecutionID()))

		result, err := e.executeNode(ctx, node, currentCtx)
		if err != nil {
			nodeStatuses[nodeID] = NodeStatusFailed

			e.emitEvent(ExecutionEvent{
				Type:      EventNodeFailed,
				NodeID:    nodeID,
				Status:    NodeStatusFailed,
				Error:     err,
				Timestamp: time.Now(),
			})

			e.logger.Error("Node execution failed",
				zap.String("node_id", nodeID),
				zap.String("execution_id", execCtx.ExecutionID()),
				zap.Error(err))

			if e.config.StopOnError {
				return nil, workflow.NewExecutionError(nodeID, "node execution failed", err)
			}
			continue
		}
		if result == nil {
			nodeStatuses[nodeID] = NodeStatusFailed
			nilResultErr := fmt.Errorf("node handler returned nil result")

			e.emitEvent(ExecutionEvent{
				Type:      EventNodeFailed,
				NodeID:    nodeID,
				Status:    NodeStatusFailed,
				Error:     nilResultErr,
				Timestamp: time.Now(),
			})

			e.logger.Error("Node execution returned nil result",
				zap.String("node_id", nodeID),
				zap.String("execution_id", execCtx.ExecutionID()))

			if e.config.StopOnError {
				return nil, workflow.NewExecutionError(nodeID, "node execution failed", nilResultErr)
			}
			continue
		}
		if result.Status == NodeStatusFailed {
			nodeStatuses[nodeID] = NodeStatusFailed
			nodeErr := result.Error
			if nodeErr == nil {
				nodeErr = fmt.Errorf("node returned failed status without error")
			}

			e.emitEvent(ExecutionEvent{
				Type:      EventNodeFailed,
				NodeID:    nodeID,
				Status:    NodeStatusFailed,
				Error:     nodeErr,
				Timestamp: time.Now(),
			})

			e.logger.Error("Node execution reported failed status",
				zap.String("node_id", nodeID),
				zap.String("execution_id", execCtx.ExecutionID()),
				zap.Error(nodeErr))

			if e.config.StopOnError {
				return nil, workflow.NewExecutionError(nodeID, "node execution failed", nodeErr)
			}
			continue
		}
		if result.Status == NodeStatusSkipped {
			skippedNodes[nodeID] = true
			nodeStatuses[nodeID] = NodeStatusSkipped
			completedCount++

			e.emitEvent(ExecutionEvent{
				Type:      EventNodeSkipped,
				NodeID:    nodeID,
				Status:    NodeStatusSkipped,
				Progress:  float64(completedCount) / float64(totalNodes),
				Timestamp: time.Now(),
			})

			e.logger.Debug("Node reported skipped status",
				zap.String("node_id", nodeID),
				zap.String("execution_id", execCtx.ExecutionID()))

			continue
		}

		// Store result in context
		currentCtx = currentCtx.SetNodeResult(nodeID, result.Output)
		executedNodes[nodeID] = true
		nodeStatuses[nodeID] = NodeStatusCompleted
		completedCount++

		e.emitEvent(ExecutionEvent{
			Type:      EventNodeCompleted,
			NodeID:    nodeID,
			Status:    NodeStatusCompleted,
			Result:    result,
			Progress:  float64(completedCount) / float64(totalNodes),
			Timestamp: time.Now(),
		})

		e.logger.Debug("Node completed",
			zap.String("node_id", nodeID),
			zap.Duration("duration", result.Duration),
			zap.String("execution_id", execCtx.ExecutionID()))
	}

	e.logger.Info("DAG execution completed",
		zap.String("execution_id", execCtx.ExecutionID()),
		zap.Int("executed", len(executedNodes)),
		zap.Int("skipped", len(skippedNodes)))

	return currentCtx, nil
}

// shouldExecuteNode determines if a node should be executed based on edge conditions.
func (e *Executor) shouldExecuteNode(
	nodeID string,
	ctx *wfcontext.ExecutionContext,
	executedNodes map[string]bool,
	skippedNodes map[string]bool,
) bool {
	incoming := e.incomingEdges[nodeID]

	// Start nodes (no incoming edges) always execute
	if len(incoming) == 0 {
		return true
	}

	// Check each incoming edge
	for _, edge := range incoming {
		sourceNode := edge.From

		// Source must have been executed (not skipped)
		if skippedNodes[sourceNode] {
			continue
		}

		if !executedNodes[sourceNode] {
			// Source not yet processed - shouldn't happen in topological order
			e.logger.Warn("Source node not yet executed",
				zap.String("source", sourceNode),
				zap.String("target", nodeID))
			continue
		}

		// Check edge condition
		if e.evaluateEdgeCondition(edge, ctx) {
			return true
		}
	}

	return false
}

// evaluateEdgeCondition evaluates an edge's condition expression.
func (e *Executor) evaluateEdgeCondition(edge *models.Edge, ctx *wfcontext.ExecutionContext) bool {
	// No condition = always true
	if edge.Condition == "" {
		return true
	}

	// Use condition evaluator if available
	if e.conditionEval != nil {
		result, err := e.conditionEval.Evaluate(edge.Condition, ctx)
		if err != nil {
			e.logger.Warn("Edge condition evaluation failed, treating as false",
				zap.String("from", edge.From),
				zap.String("to", edge.To),
				zap.String("condition", edge.Condition),
				zap.Error(err))
			return false
		}
		return result
	}

	// Default: simple boolean check from context
	// TODO: Implement Jinja2-style evaluation
	e.logger.Debug("No condition evaluator, condition treated as true",
		zap.String("condition", edge.Condition))
	return true
}

// executeNode executes a single node using the appropriate handler.
func (e *Executor) executeNode(
	ctx context.Context,
	node *models.Node,
	execCtx *wfcontext.ExecutionContext,
) (*NodeResult, error) {
	startTime := time.Now()

	handler, ok := e.handlers[node.Type]
	if !ok {
		return nil, fmt.Errorf("no handler registered for node type: %s", node.Type)
	}

	// Apply node timeout
	timeout := e.config.DefaultTimeout
	if node.Config != nil && node.Config.TimeoutSeconds > 0 {
		timeout = time.Duration(node.Config.TimeoutSeconds) * time.Second
	}

	nodeCtx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()

	// Execute via handler
	result, err := handler.HandleNode(nodeCtx, node, execCtx)
	if err != nil {
		return &NodeResult{
			NodeID:      node.ID,
			Status:      NodeStatusFailed,
			Error:       err,
			StartedAt:   startTime,
			CompletedAt: time.Now(),
			Duration:    time.Since(startTime),
		}, err
	}

	if result == nil {
		result = &NodeResult{
			NodeID: node.ID,
			Status: NodeStatusCompleted,
		}
	}

	result.NodeID = node.ID
	result.StartedAt = startTime
	result.CompletedAt = time.Now()
	result.Duration = time.Since(startTime)

	if result.Status == "" {
		result.Status = NodeStatusCompleted
	}

	return result, nil
}

// GetNextNodes returns the next nodes to execute based on edge conditions.
// Useful for async execution to determine next steps.
func (e *Executor) GetNextNodes(nodeID string, ctx *wfcontext.ExecutionContext) []string {
	outgoing := e.outgoingEdges[nodeID]
	nextNodes := make([]string, 0, len(outgoing))

	for _, edge := range outgoing {
		if e.evaluateEdgeCondition(edge, ctx) {
			nextNodes = append(nextNodes, edge.To)
		}
	}

	// Sort for deterministic order
	sort.Strings(nextNodes)

	e.logger.Debug("Determined next nodes",
		zap.String("current_node", nodeID),
		zap.Strings("next_nodes", nextNodes))

	return nextNodes
}

// emitEvent sends an execution event to the callback if registered.
func (e *Executor) emitEvent(event ExecutionEvent) {
	if e.callback != nil {
		e.callback(event)
	}
}

// DAG returns the executor's DAG.
func (e *Executor) DAG() *models.DAG {
	return e.dag
}
